File: CRA-Program/CRAMain.cpp
#include <cstring>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <limits>

using namespace std;

string locateDir() //locates the CRA-Program directory and returns it for use in the other methods
{
    string dirPath(256, '\0');
    system("sudo find $HOME | grep SummerResearch2018/CRA-Program > deleteThis.txt"); //locates all filepaths that include the CRA-Program directory and outputs to a file to read from later
    fstream f("deleteThis.txt", ios::in | ios::app); //file stream to read from deleteThis.txt
    if(f.is_open()) //this if statement confirms that the file is created, reads the filepath of the CRA-Program directory, then deletes the file used to read from
    {
        getline(f, dirPath); //the first line of the grep command is the filepath for the CRA-Program directory - all other results can be ignored
        string deleteFile(dirPath); deleteFile.append("/deleteThis.txt"); //full path of deleteThis.txt to allow for call to remove()
        remove(deleteFile.c_str()); //requires a const_ char*, which is returned by c_str()
    }
    f.close();
    return dirPath;
}

string askFile(string path) //asks for a relative filepath to an object file to be disassembled
{
    cout << "Enter the relative filepath for the object file (.obj, .o) to be disassembled. (Must be within the CRA-Program directory or any subdirectories)\n" << path << std::flush;
    char fp[40]; cin.getline(fp, 40, '\n'); path.append(fp); //takes in the rest of the filepath to the object file from the user
    return path; //gives filepath
}

string objDumpCommand(string fp) //outputs disassembly to a file in the parent directory and returns the objdump command
{
    string outfile(fp); outfile.replace(outfile.find(".", outfile.length()-6), outfile.length() ,"-dsmbl.txt"); //appends -dsmbl.txt to the name of the file given
    string command = "objdump -d -M intel "; command.append(fp); command.append(" > "); command.append(outfile); //creates the objdump command string to be passed to system() and output to file
    cout << "Attempting to output disassembly to output file within the parent directory of selected file. Output file name: " << outfile << endl; //giving user acknowledgement of input
    return command;
}

void findLEA(string path) //asks for a file to be read and outputs memory addresses of LEA commands to terminal
{
    cout << "To search for LEA instructions, input a filepath for the -dsmbl.txt file previously created, starting from the CRA-Program directory:\n" << path << std::flush;
    char fp[44]; cin.getline(fp, 44, '\n'); path.append(fp);
    fstream f(path, ios::in | ios::app);
    string instr(64, ' ');
    int ind;
    string leaFile(path, 0, path.length()-4); leaFile.append("-LEA.txt"); //stores name of output file to store all LEA instruction output in a file to reference in other functions
    fstream store(leaFile, ios::out | ios::trunc); //if file already exists, the contents need to be overwritten, hence ios::trunc as the mode for output
    store.is_open() ? cout << leaFile << " has been created and opened." << endl : cout << leaFile << " has failed to be created and opened." << endl;
    string line; //temporary string to store each line to be written to file
    while(!f.eof())
    {
        getline(f, instr);
        ind = instr.find("lea",0,3);
        if(ind != (signed)string::npos) //not sure why warning for signed comparison is necessary here but not next line vvvvvv (already converted to unsigned when assigned to int?)
            if(instr.find("leave",0,5) == string::npos)
            {
                cout << "Found LEA instruction:     " << instr << endl;
                line.assign("Found LEA instruction:    "); line.append(instr); //reset line and allow for appending the current instruction
                store << line.c_str() << endl; // formatted output to fstream store, which writes to $filename-LEA.txt
            }
    }
    flush(cout);
    store.close();
    f.close();
}

int main()
{
    //system() requires a const_ char*, which is returned by c_str() - objDumpCommand() returns the prepared objdump linux command
    system(objDumpCommand(askFile(locateDir())).c_str()); //locateDir() is called first and outputs to askFile(), then askFile accepts user input and outputs to objDumpCommand
    findLEA(locateDir()); //not working
    //cout << "Press 'R' to rerun the program." << endl;
    //char input; cin.get(input); if(input == 'r'){ flush(cout); main();} //intended to rerun the program from the start using recursion. does not work.
    cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); //pauses terminal before it closes
    return 0;
}

File: CRA-Python/CRA-Main.py
import sys
import os


def locateDir():
    """locateDir finds the directory path of CRA-Python within the directory it was downloaded to"""
    #  vv   locates all file paths that include the CRA-Python directory and outputs to a file to read from later
    os.system("find $HOME | grep SummerResearch2018/CRA-Python > deleteThis.txt")
    dirPath = " "
    try:
        file = open("deleteThis.txt", "r+", 1)  # attempts to read from the output of the directory search
        dirPath = file.readline(256)
        file.close()
        os.remove("./deleteThis.txt")
    # except(FileNotFoundError, FileExistsError, ReferenceError):
    #    print("Unable to open deleteThis.txt")
    finally:
        return dirPath[:dirPath.__len__()-1]  # get rid of newline character returned by readline


def askFile(path):
    """asks the user for a relative filepath to an object file within the CRA-Python directory to be disassembled"""
    print("Enter the relative filepath for the object file (.obj, .o) to be disassembled. ",
          "(Must be within the CRA-Python directory or any subdirectories) \n", path, "")
    path += input()
    return path


def objDumpCommand(path):
    """prepares the Linux objdump command based on that object file, then returns the command as a string"""
    outfile = path[:path.find(".")] + "-dsmbl.txt"
    print("Output of disassembly written to file within the parent directory of selected file. Output file name: ",
          outfile)
    command = "objdump -d -M intel " + path + ">" + outfile
    return command


def findLEA(path):
    """findLEA searches the disassembled file for LEA instructions and outputs each line they're found in to a new file ending in -LEA.txt. This function is
     intended to show the memory locations of buffers to potentially overflow"""
    print("To search for LEA instructions, input a relative filepath for a -dsmbl.txt: \n", path)
    path += input()
    # this is the filepath of the LEA file to be created (will be written over if it already exists
    # if it exists,that would mean the -dsmbl file was given as input in last execution
    leaFP = path[:path.find("-dsmbl")] + "-LEA.txt"
    with open(path, 'r', 1) as dsmblFile:  # opens file in the mode that overwrites the file in read/write format
        with open(leaFP, 'w+', 1) as leaFile:
            for line in dsmblFile:  # for each line read from the leaFile
                instruction = line[line.find("lea"):]
                if line.find("lea") != -1 and line.find("leave") == -1:
                    leaFile.write("Found LEA instruction:   " + instruction + "\n")  # write formatted output to leaFile
    print("The LEA instructions found in " + path + " have been written to:    " + leaFP)


os.system(""+objDumpCommand(askFile(locateDir())))
findLEA(locateDir())

# python $HOME/Desktop/ROPgadget-master/ROPgadget.py --binary $HOME/Desktop/ROPgadget-master
# /test-suite-binaries/elf-Linux-x86 >> $HOME/Desktop/ROPgadget-master/elf-Linux-x86-Output.txt
# ^^ this is command to build the chain for me, with filePath being the designated file to use as the binary
# automated chain generation is random and not targeted at all.
# refer to http://shell-storm.org/blog/Return-Oriented-Programming-and-ROPgadget-tool/ for an example
# still need to write my own outline for how I want the program to run and then link the ROP chain "manually"
# first try manually to do /bin/sh exploit then attempt automation of finding and linking ROP chain
# need to confirm whether push operations to store target memory addresses is necessary for linking
